File: usvm-ts/src/test/resources/samples/operators/Bitwise.ts
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols

class Bitwise {
    bitwiseNot(a: number): number {
        let res = ~a;

        if (a == 5) return res; // ~5 = -6
        if (a == -1) return res; // ~(-1) = 0
        if (a == 0) return res; // ~0 = -1

        return res;
    }

    bitwiseAnd(a: number, b: number): number {
        let res = a & b;

        if (a == 5 && b == 3) return res; // 5 & 3 = 1 (101 & 011 = 001)
        if (a == 15 && b == 7) return res; // 15 & 7 = 7 (1111 & 0111 = 0111)
        if (a == 0) return res; // anything & 0 = 0

        return res;
    }

    bitwiseOr(a: number, b: number): number {
        let res = a | b;

        if (a == 5 && b == 3) return res; // 5 | 3 = 7 (101 | 011 = 111)
        if (a == 0 && b == 0) return res; // 0 | 0 = 0
        if (a == 15 && b == 16) return res; // 15 | 16 = 31

        return res;
    }

    bitwiseXor(a: number, b: number): number {
        let res = a ^ b;

        if (a == 5 && b == 3) return res; // 5 ^ 3 = 6 (101 ^ 011 = 110)
        if (a == 7 && b == 7) return res; // x ^ x = 0
        if (a == 0) return res; // 0 ^ x = x

        return res;
    }

    leftShift(a: number, b: number): number {
        let res = a << b;

        if (a == 5 && b == 1) return res; // 5 << 1 = 10
        if (a == 1 && b == 3) return res; // 1 << 3 = 8
        if (b == 0) return res; // x << 0 = x

        return res; // fallback case
    }

    rightShift(a: number, b: number): number {
        let res = a >> b;

        if (a == 10 && b == 1) return res; // 10 >> 1 = 5
        if (a == -8 && b == 2) return res; // -8 >> 2 = -2 (sign extends)
        if (b == 0) return res; // x >> 0 = x

        return res; // fallback case
    }

    unsignedRightShift(a: number, b: number): number {
        let res = a >>> b;

        if (a == 10 && b == 1) return res; // 10 >>> 1 = 5
        if (a == -1 && b == 1) return res; // -1 >>> 1 = 2147483647 (max positive int32)
        if (b == 0) return res; // x >>> 0 = x

        return res; // fallback case
    }
}

File: usvm-ts/src/test/kotlin/org/usvm/samples/operators/Bitwise.kt
package org.usvm.samples.operators

import org.jacodb.ets.model.EtsScene
import org.usvm.api.TsTestValue
import org.usvm.util.TsMethodTestRunner
import org.usvm.util.eq
import kotlin.test.Test

class Bitwise : TsMethodTestRunner() {
    private val tsPath = "/samples/operators/Bitwise.ts"

    override val scene: EtsScene = loadScene(tsPath)

    @Test
    fun `bitwise NOT`() {
        val method = getMethod("bitwiseNot")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, r ->
                // ~5 = -6
                (a eq 5) && (r eq -6)
            },
            { a, r ->
                // ~(-1) = 0
                (a eq -1) && (r eq 0)
            },
            { a, r ->
                // ~0 = -1
                (a eq 0) && (r eq -1)
            },
            invariants = arrayOf(
                // TODO: implement TS behaviour for bitwise NOT
                { _, r -> true }
            )
        )
    }

    @Test
    fun `bitwise AND`() {
        val method = getMethod("bitwiseAnd")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, b, r ->
                // 5 & 3 = 1
                (a eq 5) && (b eq 3) && (r eq 1)
            },
            { a, b, r ->
                // 15 & 7 = 7
                (a eq 15) && (b eq 7) && (r eq 7)
            },
            { a, b, r ->
                // 0 & b = 0
                (a eq 0) && (r eq 0)
            },
            invariants = arrayOf(
                // TODO: implement TS behaviour for bitwise AND
                { _, _, r -> true }
            )
        )
    }

    @Test
    fun `bitwise OR`() {
        val method = getMethod("bitwiseOr")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, b, r ->
                // 5 | 3 = 7
                (a eq 5) && (b eq 3) && (r eq 7)
            },
            { a, b, r ->
                // 0 | 0 = 0
                (a eq 0) && (b eq 0) && (r eq 0)
            },
            { a, b, r ->
                // 15 | 16 = 31
                (a eq 15) && (b eq 16) && (r eq 31)
            },
            invariants = arrayOf(
                // TODO: implement TS behaviour for bitwise OR
                { _, _, r -> true }
            )
        )
    }

    @Test
    fun `bitwise XOR`() {
        val method = getMethod("bitwiseXor")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, b, r ->
                // 5 ^ 3 = 6
                (a eq 5) && (b eq 3) && (r eq 6)
            },
            { a, b, r ->
                // 7 ^ 7 = 0
                (a eq 7) && (b eq 7) && (r eq 0)
            },
            { a, b, r ->
                // 0 ^ b = b
                (a eq 0) && (r == b)
            },
            invariants = arrayOf(
                // TODO: implement TS behaviour for bitwise XOR
                { _, _, r -> true }
            )
        )
    }

    @Test
    fun `left shift`() {
        val method = getMethod("leftShift")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, b, r ->
                // 5 << 1 = 10
                (a eq 5) && (b eq 1) && (r eq 10)
            },
            { a, b, r ->
                // 1 << 3 = 8
                (a eq 1) && (b eq 3) && (r eq 8)
            },
            { a, b, r ->
                // a << 0 = a
                (b eq 0) && (r == a)
            },
            invariants = arrayOf(
                // TODO: implement TS behaviour for left shift
                { _, _, r -> true }
            )
        )
    }

    @Test
    fun `right shift`() {
        val method = getMethod("rightShift")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, b, r ->
                // 10 >> 1 = 5
                (a eq 10) && (b eq 1) && (r eq 5)
            },
            { a, b, r ->
                // -8 >> 2 = -2
                (a eq -8) && (b eq 2) && (r eq -2)
            },
            { a, b, r ->
                // a >> 0 = a
                (b eq 0) && (r eq a)
            },
            invariants = arrayOf(
                // TODO: implement TS behaviour for right shift
                { _, _, r -> true }
            )
        )
    }

    @Test
    fun `unsigned right shift`() {
        val method = getMethod("unsignedRightShift")
        discoverProperties<TsTestValue.TsNumber, TsTestValue.TsNumber, TsTestValue.TsNumber>(
            method = method,
            { a, b, r ->
                // 10 >>> 1 = 5
                (a eq 10) && (b eq 1) && (r eq 5)
            },
            { a, b, r ->
                // -1 >>> 1 = 2147483647
                (a eq -1) && (b eq 1) && (r eq 2147483647)
            },
            { a, b, r ->
                // x >>> 0 = x
                (b eq 0) && (r eq a)
            },
            invariants = arrayOf(
                // TODO: implement TS behaviour for unsigned right shift
                { _, _, r -> true }
            )
        )
    }
}
